src/webassembly/mod.rs

```diff
@@ -85,17 +87,128 @@ pub fn run_instance(
     module: &Module,
     instance: &mut Instance,
     _path: &str,
-    _args: Vec<&str>,
+    args: Vec<&str>,
 ) -> CallResult<()> {
     let main_name = if is_emscripten_module(module) {
         "_main"
     } else {
         "main"
     };
 
-    // TODO handle args
-    instance.call(main_name, &[])?;
+    // Get main arguments.
+    let main_args = get_main_args(main_name, args, instance).unwrap();
+
+    // Call main function with the arguments.
+    instance.call(main_name, &main_args)?;
+
     // TODO atinit and atexit for emscripten
 
     Ok(())
 }
+
+/// Passes arguments from the host to the WebAssembly instance.
+fn get_main_args(
+    main_name: &str,
+    args: Vec<&str>,
+    instance: &mut Instance,
+) -> CallResult<Vec<Value>> {
+    // Getting main function signature.
+    let func_sig = instance.get_signature(main_name)?;
+    let params = &func_sig.params;
+
+    // Check for a () or (i32, i32) sig.
+    match params.as_slice() {
+        &[Type::I32, Type::I32] => {
+            // Copy strings into wasm memory and get addresses to them.
+            let string_addresses = args
+                .iter()
+                .map(|string| copy_string_into_wasm(instance, (*string).to_string()).unwrap())
+                .collect();
+
+            // Create a wasm array to the strings.
+            let array = create_wasm_array(instance, string_addresses).unwrap();
+
+            Ok(vec![
+                Value::I32(array as i32),
+                Value::I32(args.len() as i32),
+            ])
+        }
+        &[] => Ok(vec![]),
+        _ => Err(CallError::Signature {
+            expected: FuncSig {
+                params: vec![Type::I32, Type::I32],
+                returns: vec![],
+            },
+            found: params.to_vec(),
+        }
+        .into()),
+    }
+}
+
+/// Copy rust string to wasm instance.
+fn copy_string_into_wasm(instance: &mut Instance, string: String) -> CallResult<u32> {
+    let string_len = string.len();
+
+    let space_offset = instance
+        .call("_malloc", &[Value::I32((string_len as i32) + 1)])
+        .unwrap();
+
+    let space_offset = match space_offset.as_slice() {
+        &[Value::I32(res)] => Some(res as u32),
+        _ => None,
+    }.unwrap();
+
+    let raw_memory = instance.inner.vmctx.memory(0)[space_offset as usize] as *mut u8;
```

Comment:
Rather than doing this, I think would be better to have a `vmctx` function in the instance itself (that points to the inner vmctx).

This way, you can do `instance.vmctx().memory(0)` ...

Reply:
cc: @lachlansneff ?

Reply:
I added a `ctx()` function in the PR.